pkg/matrix/matrix.go

```diff
@@ -0,0 +1,36 @@
+package matrix
+
+import (
+	"errors"
+)
+
+func DotProduct(m1 [][]float64, m2 [][]float64) ([][]float64, error) {
+	if len(m1[0]) != len(m2) {
+		return nil, errors.New("the number of columns in the first matrix must be equal to the number of rows in the second matrix")
```

Comment:
You can calculate both matrices shapes at the beginning of the function. This way you don't need to do `len(m1[0])`, `len(m2)`, `len(m1)`, `len(m2[i])`, etc. Depending on the context, each `len` call is `O(N)`. I'm not entirely sure if this is the case right now, but if yes it'd mean that you are adding much more complexity to the function. If you are curious about Go's behavior you can:
- Read more about the `len` function and what cases it recalculates the array size.
- Do a time comparison for larger matrices.